html tag parse fix

elliefm opened the pull request:
This tests and fixes (I think) a bug @robmueller and @wolfsage tracked down from, where the html_attr state wasn't being reset between tags (there's a big writeup in Linear).

I don't know lib/charset especially well, so my confidence isn't high that I haven't missed a detail...

I found a couple of surprising-but-unrelated things while in here, so I've added failing cunit tests for them (as WIP commits).  Do we want to address those too?  If not, I'll just drop those commits.  In the meantime, CI will fail on this PR because of those failing tests...

@rsto I've marked this as a draft for now until we know what (if anything) we want to do with the two WIP commits, but please review now anyway.

rsto reviewed (commented):
Thanks a lot, the bug fix commit is great as-is and I already have cherry-picked it on the feature branch. For the other two please see my comments in the review.

rsto commented on cunit/charset.testc:
```diff
@@ -1213,6 +1213,10 @@ static void test_extract(void)
+    // XXX multiple hrefs in same A tag (invalid!)
+    TESTCASE("<a href=\"https://foo/bar\" href=\"https://baz/qum\"/>",
+             "us-ascii", ENCODING_NONE, "HTML",
+             "what do we want here?");
```
This is required to make the HTML extractor use the last value (which also solve the same issue for "alt"):

     diff --git a/lib/charset.c b/lib/charset.c
     index d3034b6b3c..8704ef0f77 100644
     --- a/lib/charset.c
     +++ b/lib/charset.c
     @@ -1323,6 +1323,10 @@ static void html_attr_start(struct striphtml_state *s)
                  s->attr.typ = HATTR_ALT;
              }
          }
     +
     +    if (s->attr.typ != HATTR_NONE) {
     +        dynarray_truncate(&s->attr.vals[s->attr.typ], 0);
     +    }
      }

      static void html_attr_putc(struct striphtml_state *s, uint32_t c)

I say we just do this, rather than supporting the edge case of finding invalid html.

elliefm commented:
@rsto I've made your suggested change to use the last attr value, and updated and extended the tests for it.  I've also tidied up the comment about only using img alt if src is present and removed the failing test, and squashed everything into a single sane commit.

This PR will fail CI at the moment because JMAPTestSuite has been updated, but the base branch doesn't have the updates to accommodate it yet.  I expect it to fail these tests specifically:

```
[FAILED] Cyrus::JMAPTestSuite.t:Email:import:one-fails-another-succeeds
[FAILED] Cyrus::JMAPTestSuite.t:Email:import:good-imports
[FAILED] Cyrus::JMAPTestSuiteWS.t:Email:import:one-fails-another-succeeds
[FAILED] Cyrus::JMAPTestSuiteWS.t:Email:import:good-imports
```

Once approved, I'll "Rebase and merge" this one rather than the usual "Merge pull request", so that the charset_extract_html_urls branch doesn't become cluttered with a merge commit.  Sound good?

rsto reviewed (approved):
Many thanks for this.
